### DISCUSSION TITLE
Mixer and midi improvements

### DISCUSSION THREAD

**Discussion by pumaelbusta** [hydrogen-music/hydrogen#1120] [Ideas]
Hi, 
I wanted to assign midi controllers to instrument volume controls as it's visible in the Hydrogen Mixer. I was a bit surprised I could assign a controller to the main volume only.
I took a look at the code and I see that the mapping works with hard-coded stuff, e.g. MASTER_VOLUME_ABSOLUTE .
Would it be possbile to add a control to each instrument instance with a dynamic name, like VOLUME_$(LABEL)?

p.s.: the mixer instrument label says "double click to edit" but it doesn't work for me and actually any of the 3 mouse buttons does the same - only adjusting the volume (it's version 1.0.0-rc1 installed from ubuntu focal repo).

An midi learn style controller assignment would be extremely helpful, I don't know how hard is that to implement.

Are there plans for such improvements? Unfortunately I'm not a programmer so I can't contribute in these.

**Comment by theGreatWhiteShark** [comment]
> Would it be possbile to add a control to each instrument instance with a dynamic name, like VOLUME_$(LABEL)?

There is a `STRIP_VOLUME_ABSOLUTE` with which you should be able to control the volume of the different instrument/strips within the mixer.

> p.s.: the mixer instrument label says "double click to edit" but it doesn't work for me and actually any of the 3 mouse buttons does the same - only adjusting the volume (it's version 1.0.0-rc1 installed from ubuntu focal repo).

Thanks for reporting. This tooltip is set for the name of the strip. But, indeed, it can not be altered. I will remove it.

> An midi learn style controller assignment would be extremely helpful, I don't know how hard is that to implement.

So, there are quite a number of plans already. But none of them involve MIDI. But let's see. Maybe someone is interested in the topic and creates a patch. Or you decide to give both programming and this feature a try :wink:

**Comment by theGreatWhiteShark** [comment]
> An midi learn style controller assignment would be extremely helpful, I don't know how hard is that to implement.

Just found the following in the code: if you press [shift] + left mouse key you can make a widget - like a fader, button, or knob - MIDI learnable.

**Reply by pumaelbusta** [reply]
Wow, it works, thank you very much. I couldn't imagine how it'd work with STRIP_VOLUME_ABSOLUTE as I woudln't be able to tell which fader or knob I want to assign. I'll try to look up if it's documented anywhere and spread the word. :)

**Reply by theGreatWhiteShark** [reply]
I never tried it but I would expect it to require a number indicating the position of the strip in the mixer (start counting at 1; from left to right).
